File: app/frontend/src/lib/workspaceRuntime.ts
import { buildAuthHeaders } from "./authToken";

export interface WorkspacePreviewBundle {
  project_id?: number;
  status?: string;
  preview_session_key: string;
  preview_expires_at?: string | null;
  preview_frontend_url: string;
  preview_backend_url: string;
  frontend_port?: number | null;
  backend_port?: number | null;
  frontend_status?: string | null;
  backend_status?: string | null;
}


export function buildPreviewUrl(previewUrl: string): string {
  return previewUrl;
}

export async function ensureWorkspaceRuntime(projectId: number): Promise<WorkspacePreviewBundle> {
  const response = await fetch(`/api/v1/workspace-runtime/projects/${projectId}/ensure`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...buildAuthHeaders(),
    },
  });
  if (!response.ok) {
    throw new Error(`Failed to ensure runtime: ${response.status}`);
  }
  return response.json();
}

File: app/backend/schemas/workspace_runtime.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class WorkspaceRuntimeStatusResponse(BaseModel):
    project_id: int
    status: str
    container_name: str
    preview_session_key: str
    preview_expires_at: Optional[datetime] = None
    preview_frontend_url: str
    preview_backend_url: str
    frontend_port: Optional[int] = None
    backend_port: Optional[int] = None
    frontend_status: Optional[str] = None
    backend_status: Optional[str] = None
